File: engine/src/main/java/io/apitomy/flow/engine/ParallelRegions.java
package io.apitomy.flow.engine;

import io.apitomy.flow.model.NodeType;
import io.apitomy.flow.model.Workflow;
import io.apitomy.flow.model.WorkflowEdge;
import io.apitomy.flow.model.WorkflowNode;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Static structural analysis of a workflow's parallel regions. Classifies fork nodes (all-unconditional
 * multi-out), pairs each fork with its synchronizing join, and reports structural problems that make a
 * region ill-formed. Pure and side-effect free; shared by the validator and the engine so both agree on
 * what a fork is, what its join is, and what a well-formed region looks like.
 */
public final class ParallelRegions {

    /** A structural problem discovered during analysis. */
    public record Problem(String code, String nodeId) {}

    private final Set<String> forks;
    private final Map<String, String> forkToJoin;
    private final Set<String> joins;
    private final Map<String, Set<String>> joinIncomingEdgeIds;
    private final List<Problem> problems;

    private ParallelRegions(Set<String> forks, Map<String, String> forkToJoin, Set<String> joins,
                            Map<String, Set<String>> joinIncomingEdgeIds, List<Problem> problems) {
        this.forks = forks;
        this.forkToJoin = forkToJoin;
        this.joins = joins;
        this.joinIncomingEdgeIds = joinIncomingEdgeIds;
        this.problems = problems;
    }

    /**
     * Analyzes a workflow's parallel structure.
     *
     * @param workflow the workflow to analyze
     * @return the computed regions and any structural problems
     */
    public static ParallelRegions analyze(Workflow workflow) {
        Set<String> forks = new LinkedHashSet<>();
        Map<String, String> forkToJoin = new LinkedHashMap<>();
        Set<String> joins = new LinkedHashSet<>();
        Map<String, Set<String>> joinIncoming = new LinkedHashMap<>();
        List<Problem> problems = new ArrayList<>();

        for (WorkflowNode node : workflow.nodes()) {
            List<WorkflowEdge> outgoing = workflow.getOutgoingEdges(node.id());
            if (outgoing.size() < 2) {
                continue;
            }
            boolean anyUnconditionalNonDefault = outgoing.stream()
                .anyMatch(e -> isUnconditional(e) && !e.isDefault());
            boolean allUnconditionalNonDefault = outgoing.stream()
                .allMatch(e -> isUnconditional(e) && !e.isDefault());

            if (allUnconditionalNonDefault) {
                forks.add(node.id());
            } else if (anyUnconditionalNonDefault) {
                // mixes fork-shaped edges with conditional/default edges — ambiguous
                problems.add(new Problem("MIXED_FORK_EDGES", node.id()));
            }
        }

        for (String forkId : forks) {
            String join = findJoin(workflow, forkId, problems);
            if (join != null) {
                forkToJoin.put(forkId, join);
                joins.add(join);
                Set<String> incoming = new LinkedHashSet<>();
                for (WorkflowEdge e : workflow.getIncomingEdges(join)) {
                    incoming.add(e.id());
                }
                joinIncoming.put(join, incoming);
            }
        }

        return new ParallelRegions(forks, forkToJoin, joins, joinIncoming, problems);
    }

    private static boolean isUnconditional(WorkflowEdge e) {
        return e.condition() == null || e.condition().isBlank();
    }

    /**
     * Finds the synchronizing join for a fork: the first node where every branch leaving the fork
     * re-converges. Adds a {@code FORK_WITHOUT_JOIN} / {@code PARALLEL_BRANCH_REACHES_END} /
     * {@code UNBALANCED_PARALLEL} problem when no single balanced convergence node exists.
     */
    private static String findJoin(Workflow workflow, String forkId, List<Problem> problems) {
        List<WorkflowEdge> branches = workflow.getOutgoingEdges(forkId);
        List<Set<String>> reachablePerBranch = new ArrayList<>();
        boolean anyBranchReachesEnd = false;

        for (WorkflowEdge branch : branches) {
            Set<String> reachable = new LinkedHashSet<>();
            Deque<String> queue = new ArrayDeque<>();
            queue.add(branch.target());
            while (!queue.isEmpty()) {
                String current = queue.poll();
                if (!reachable.add(current)) {
                    continue;
                }
                WorkflowNode node = workflow.findNodeById(current).orElse(null);
                if (node != null && node.type() == NodeType.END) {
                    anyBranchReachesEnd = true;
                }
                for (WorkflowEdge out : workflow.getOutgoingEdges(current)) {
                    queue.add(out.target());
                }
            }
            reachablePerBranch.add(reachable);
        }

        // The join is the earliest node reachable from ALL branches.
        Set<String> common = new LinkedHashSet<>(reachablePerBranch.get(0));
        for (int i = 1; i < reachablePerBranch.size(); i++) {
            common.retainAll(reachablePerBranch.get(i));
        }
        if (common.isEmpty()) {
            problems.add(new Problem(anyBranchReachesEnd ? "PARALLEL_BRANCH_REACHES_END"
                : "FORK_WITHOUT_JOIN", forkId));
            return null;
        }

        // Earliest common node = the one whose incoming edges come from every branch. Choose the common
        // node reachable in the fewest steps from the fork along any branch (BFS order preserved above).
        String join = firstCommon(reachablePerBranch, common);

        // Balance check: every branch must reach the join without first hitting END.
        for (WorkflowEdge branch : branches) {
            if (reachesEndBeforeJoin(workflow, branch.target(), join)) {
                problems.add(new Problem("PARALLEL_BRANCH_REACHES_END", forkId));
                return null;
            }
        }
        return join;
    }

    private static String firstCommon(List<Set<String>> reachablePerBranch, Set<String> common) {
        // Use the first branch's insertion order (BFS) as the canonical ordering and return the first
        // element that is common to all branches.
        for (String candidate : reachablePerBranch.get(0)) {
            if (common.contains(candidate)) {
                return candidate;
            }
        }
        return common.iterator().next();
    }

    private static boolean reachesEndBeforeJoin(Workflow workflow, String start, String join) {
        Set<String> visited = new HashSet<>();
        Deque<String> queue = new ArrayDeque<>();
        queue.add(start);
        while (!queue.isEmpty()) {
            String current = queue.poll();
            if (current.equals(join) || !visited.add(current)) {
                continue;
            }
            WorkflowNode node = workflow.findNodeById(current).orElse(null);
            if (node != null && node.type() == NodeType.END) {
                return true;
            }
            for (WorkflowEdge out : workflow.getOutgoingEdges(current)) {
                queue.add(out.target());
            }
        }
        return false;
    }

    /** @return true if the node is a fork (all-unconditional, ≥2 outgoing edges). */
    public boolean isFork(String nodeId) { return forks.contains(nodeId); }

    /** @return true if the node is the synchronizing join of some fork. */
    public boolean isJoin(String nodeId) { return joins.contains(nodeId); }

    /** @return the join node id paired with the given fork, or {@code null} if none. */
    public String joinFor(String forkNodeId) { return forkToJoin.get(forkNodeId); }

    /** @return the incoming edge ids a join must collect before it fires. */
    public Set<String> incomingEdgeIds(String joinNodeId) {
        return joinIncomingEdgeIds.getOrDefault(joinNodeId, Set.of());
    }

    /** @return the structural problems discovered during analysis. */
    public List<Problem> problems() { return problems; }
}

File: ui/src/simulation/parallelRegions.ts
import { type Workflow, type WorkflowEdge } from '../types/workflow.ts';

/** A structural problem discovered during parallel-region analysis. */
export interface ParallelProblem {
    code: string;
    nodeId: string;
}

/**
 * Result of static parallel-region analysis: fork/join classification plus any structural problems.
 * Mirrors the Java {@code io.apitomy.flow.engine.ParallelRegions} so the validator and the simulation
 * agree on what a fork is, what its join is, and what a well-formed region looks like. Pure and
 * side-effect free.
 */
export interface ParallelAnalysis {
    /** @returns true if the node is a fork (≥2 outgoing edges, all unconditional and non-default). */
    isFork(nodeId: string): boolean;
    /** @returns true if the node is the synchronizing join of some fork. */
    isJoin(nodeId: string): boolean;
    /** @returns the join node id paired with the given fork, or undefined if none. */
    joinFor(forkNodeId: string): string | undefined;
    /** @returns a fresh copy of the incoming edge ids a join must collect before it fires. */
    incomingEdgeIds(joinNodeId: string): Set<string>;
    /** The structural problems discovered during analysis. */
    problems: ParallelProblem[];
}

function isUnconditional(e: WorkflowEdge): boolean {
    return !e.condition || e.condition.trim() === '';
}

function outgoing(workflow: Workflow, nodeId: string): WorkflowEdge[] {
    return workflow.edges.filter(e => e.source === nodeId);
}

function isEndNode(workflow: Workflow, nodeId: string): boolean {
    return workflow.nodes.find(n => n.id === nodeId)?.type === 'end';
}

/**
 * Analyzes a workflow's parallel structure.
 *
 * @param workflow the workflow to analyze
 * @returns the computed regions and any structural problems
 */
export function analyzeParallelRegions(workflow: Workflow): ParallelAnalysis {
    const forks = new Set<string>();
    const joins = new Set<string>();
    const forkToJoin = new Map<string, string>();
    const joinIncoming = new Map<string, Set<string>>();
    const problems: ParallelProblem[] = [];

    for (const node of workflow.nodes) {
        const out = outgoing(workflow, node.id);
        if (out.length < 2) {
            continue;
        }
        const anyForkShaped = out.some(e => isUnconditional(e) && !e.isDefault);
        const allForkShaped = out.every(e => isUnconditional(e) && !e.isDefault);
        if (allForkShaped) {
            forks.add(node.id);
        } else if (anyForkShaped) {
            problems.push({ code: 'MIXED_FORK_EDGES', nodeId: node.id });
        }
    }

    for (const forkId of forks) {
        const join = findJoin(workflow, forkId, problems);
        if (join !== null) {
            forkToJoin.set(forkId, join);
            joins.add(join);
            const incoming = new Set<string>();
            for (const e of workflow.edges) {
                if (e.target === join) {
                    incoming.add(e.id);
                }
            }
            joinIncoming.set(join, incoming);
        }
    }

    return {
        isFork: (nodeId) => forks.has(nodeId),
        isJoin: (nodeId) => joins.has(nodeId),
        joinFor: (forkNodeId) => forkToJoin.get(forkNodeId),
        incomingEdgeIds: (joinNodeId) => new Set(joinIncoming.get(joinNodeId) ?? []),
        problems,
    };
}

/**
 * Finds the synchronizing join for a fork: the earliest node where every branch leaving the fork
 * re-converges. Records FORK_WITHOUT_JOIN / PARALLEL_BRANCH_REACHES_END when no single balanced
 * convergence node exists. Mirrors {@code ParallelRegions.findJoin}.
 */
function findJoin(workflow: Workflow, forkId: string, problems: ParallelProblem[]): string | null {
    const branches = outgoing(workflow, forkId);
    const reachablePerBranch: Set<string>[] = [];
    let anyBranchReachesEnd = false;

    for (const branch of branches) {
        const reachable = new Set<string>();
        const queue: string[] = [branch.target];
        while (queue.length > 0) {
            const current = queue.shift() as string;
            if (reachable.has(current)) {
                continue;
            }
            reachable.add(current);
            if (isEndNode(workflow, current)) {
                anyBranchReachesEnd = true;
            }
            for (const out of outgoing(workflow, current)) {
                queue.push(out.target);
            }
        }
        reachablePerBranch.push(reachable);
    }

    // The join is the earliest node reachable from ALL branches.
    let common = new Set<string>(reachablePerBranch[0]);
    for (let i = 1; i < reachablePerBranch.length; i++) {
        common = new Set([...common].filter(n => reachablePerBranch[i].has(n)));
    }
    if (common.size === 0) {
        problems.push({
            code: anyBranchReachesEnd ? 'PARALLEL_BRANCH_REACHES_END' : 'FORK_WITHOUT_JOIN',
            nodeId: forkId,
        });
        return null;
    }

    // Earliest common node, using branch[0]'s BFS insertion order as the canonical ordering.
    let join: string | null = null;
    for (const candidate of reachablePerBranch[0]) {
        if (common.has(candidate)) {
            join = candidate;
            break;
        }
    }
    if (join === null) {
        join = common.values().next().value as string;
    }

    // Balance: every branch must reach the join without first hitting END.
    for (const branch of branches) {
        if (reachesEndBeforeJoin(workflow, branch.target, join)) {
            problems.push({ code: 'PARALLEL_BRANCH_REACHES_END', nodeId: forkId });
            return null;
        }
    }
    return join;
}

function reachesEndBeforeJoin(workflow: Workflow, start: string, join: string): boolean {
    const visited = new Set<string>();
    const queue: string[] = [start];
    while (queue.length > 0) {
        const current = queue.shift() as string;
        if (current === join || visited.has(current)) {
            continue;
        }
        visited.add(current);
        if (isEndNode(workflow, current)) {
            return true;
        }
        for (const out of outgoing(workflow, current)) {
            queue.push(out.target);
        }
    }
    return false;
}
